File: MATLAB/GenerateSummaryStatistics.m
function sx = GenerateSummaryStatistics(SummaryStatData, CellTracking, Xmax)  
    
    Nred = mean(SummaryStatData.Nred_ss);
    Nyellow = mean(SummaryStatData.Nyellow_ss);
    Ngreen = mean(SummaryStatData.Ngreen_ss);


    if CellTracking
        red_distance = SummaryStatData.red_distance;
        yellow_distance = SummaryStatData.yellow_distance;
        green_distance = SummaryStatData.green_distance;
        red_time = SummaryStatData.red_time;
        yellow_time = SummaryStatData.yellow_time;
        green_time = SummaryStatData.green_time;

        %red_distance(red_time == 0) = NaN; %by default red is always recorded
        yellow_distance(yellow_time == 0) = NaN;
        green_distance(green_time == 0) = NaN;

        red_distance_avg = mean(red_distance,'all','omitnan');
        yellow_distance_avg = mean(yellow_distance,'all','omitnan');
        green_distance_avg = mean(green_distance,'all','omitnan');

        sx = [Nred, Nyellow, Ngreen, red_distance_avg, yellow_distance_avg, green_distance_avg];
    else
        red_position = SummaryStatData.red_position;
        yellow_position = SummaryStatData.yellow_position;
        green_position = SummaryStatData.green_position;
        
        red_distance_med = [median(red_position{1}(red_position{1} <= Xmax/2)),median(red_position{1}(red_position{1} > Xmax/2))];
        yellow_distance_med = [median(yellow_position{1}(yellow_position{1} <= Xmax/2)),median(yellow_position{1}(yellow_position{1} > Xmax/2))]; 
        green_distance_med = [median(green_position{1}(green_position{1} <= Xmax/2)),median(green_position{1}(green_position{1} > Xmax/2))]; 
        
        red_distance_iqr = [iqr(red_position{1}(red_position{1} <= Xmax/2)),iqr(red_position{1}(red_position{1} > Xmax/2))];
        yellow_distance_iqr = [iqr(yellow_position{1}(yellow_position{1} <= Xmax/2)),iqr(yellow_position{1}(yellow_position{1} > Xmax/2))]; 
        green_distance_iqr = [iqr(green_position{1}(green_position{1} <= Xmax/2)),iqr(green_position{1}(green_position{1} > Xmax/2))]; 
        
        sx = [Nred, Nyellow, Ngreen, red_distance_med, yellow_distance_med, green_distance_med, red_distance_iqr, yellow_distance_iqr, green_distance_iqr];
        
    end
end

File: R/GenerateSummaryStatistics.R
GenerateSummaryStatistics <- function(data, CellTracking = TRUE, Xmax){
  
  if (CellTracking){
    
    # extract variables from list
    red_distance <- data$red_distance
    yellow_distance <- data$yellow_distance
    green_distance <- data$green_distance
    red_time <- data$red_time
    yellow_time <- data$yellow_time
    green_time <- data$green_time
    
    #remove observations where phases werent reached
    #red_distance[red_time == 0] <- NA - by default red is always recorded
    yellow_distance[yellow_time <= 0] <- NA 
    green_distance[green_time <= 0] <- NA
    
    red_distance_avg <- mean(red_distance, na.rm = T) #compute average distance traveled
    yellow_distance_avg <- mean(yellow_distance, na.rm = T) #compute average distance traveled
    green_distance_avg <- mean(green_distance, na.rm = T) #compute average distance traveled
  
    return(c(mean(data$Nred), mean(data$Nyellow), mean(data$Ngreen), red_distance_avg, yellow_distance_avg, green_distance_avg)) 
    
  } else { #spatio temporal data
    
    #extract variables from list
    red_position <- data$red_position
    yellow_position <- data$yellow_position
    green_position <- data$green_position
    
    #calculate median position on left and right sides
    red_distance_med <- apply(matrix(unlist(lapply(red_position, FUN = function(x){c(median(x[x <= Xmax/2], na.rm = T),median(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    yellow_distance_med <- apply(matrix(unlist(lapply(yellow_position, FUN = function(x){c(median(x[x <= Xmax/2], na.rm = T),median(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    green_distance_med <- apply(matrix(unlist(lapply(green_position, FUN = function(x){c(median(x[x <= Xmax/2], na.rm = T),median(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    
    #calculate interquartile range on left and right sides
    red_distance_iqr <- apply(matrix(unlist(lapply(red_position, FUN = function(x){c(IQR(x[x <= Xmax/2], na.rm = T),IQR(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    yellow_distance_iqr <- apply(matrix(unlist(lapply(yellow_position, FUN = function(x){c(IQR(x[x <= Xmax/2], na.rm = T),IQR(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    green_distance_iqr <- apply(matrix(unlist(lapply(green_position, FUN = function(x){c(IQR(x[x <= Xmax/2], na.rm = T),IQR(x[x > Xmax/2], na.rm = T))})), ncol = 2, byrow = T), MARGIN = 2, FUN = function(x){mean(x, na.rm = T)})
    
    return(c(mean(data$Nred), mean(data$Nyellow), mean(data$Ngreen), red_distance_med, yellow_distance_med, green_distance_med, red_distance_iqr, yellow_distance_iqr, green_distance_iqr))
  }
}
